### PR TITLE
Implementation of $compute

### PR DESCRIPTION
**Pull request by mtroth-microsoft**
### Issues
This pull request fixes issue #799 .

### Description
Adding support to ODL parser for the $compute query option.

### Checklist (Uncheck if it is not completed)
- [ x ] Test cases added
- [ x ] Build and test with one-click build and test script passed

### CODE REVIEW

**Comment by kosinsky** [comment]
Could you support inside $apply too? For example, GET ~/Sales?$apply=compute(Amount mul Product/TaxRate as Tax) I think it should be the same approach as $filter and filter transformation in $apply
http://docs.oasis-open.org/odata/odata-data-aggregation-ext/v4.0/cs02/odata-data-aggregation-ext-v4.0-cs02.html#_Toc435016588

File: src/Microsoft.OData.Core/UriParser/Parsers/UriQueryExpressionParser.cs [hunk 1]
```diff
@@ -205,6 +205,38 @@ internal static string GetEdmConstantNames(IEdmTypeReference edmTypeReference)
...
+            {
+                return transformationTokens;
+            }
+
+            this.recursionDepth = 0;
+            this.lexer = CreateLexerForFilterOrOrderByOrApplyExpression(compute);
+
+            while (true)
+            {
+                ComputeToken computed = this.ParseComputeExpression();
```
Comment (robward-ms) [inline src/Microsoft.OData.Core/UriParser/Parsers/UriQueryExpressionParser.cs]: >computed [](start = 29, length = 8)

could computed by null or invalid?

File: src/Microsoft.OData.Core/UriParser/SyntacticAst/ComputeToken.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            this.expression = expression;
+            this.alias = alias;
+        }
+
+        /// <summary>
+        /// Gets the kind of this token.
+        /// </summary>
+        public override QueryTokenKind Kind
+        {
+            get { return QueryTokenKind.Compute; }
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/UriParser/SyntacticAst/ComputeToken.cs]: format nit: the bracelets are on the line as the return, whereas they're on separate lines for the other properties below

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ODataUriParserTests.cs [hunk 3]
```diff
@@ -677,7 +677,152 @@ public void ParsePathFunctionWithNullEntitySetPath()
             pathSegmentList[0].ShouldBeOperationImportSegment(getCurrentCustomerImport);
             pathSegmentList[1].ShouldBeNavigationPropertySegment(customerDetail);
         }
+        #endregion
 
+        #region Parse Compute Tests
```
Comment (robward-ms) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ODataUriParserTests.cs]: >Tests [](start = 30, length = 5)

What about negative testing? Maybe a test for missing "as"?

**Comment by mtroth-microsoft** [comment]
As discussed offline, this makes sense but will be handled through a separate PR.

---
In reply to: [309063099](https://github.com/OData/odata.net/pull/827#issuecomment-309063099) [](ancestors = 309063099)

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ODataUriParserTests.cs (hunk 3 above)
Comment (mtroth-microsoft) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ODataUriParserTests.cs]: I'll add something for 'as', but the remainder should be covered by the expression parsing reuse.

---
In reply to: [123090919](https://github.com/OData/odata.net/pull/827#discussion_r123090919) [](ancestors = 123090919)

File: src/Microsoft.OData.Core/UriParser/Parsers/UriQueryExpressionParser.cs (hunk 1 above)
Comment (mtroth-microsoft) [inline src/Microsoft.OData.Core/UriParser/Parsers/UriQueryExpressionParser.cs]: computed can't be null since we are testing for the a null compute string and returning empty token list if encountered. computed can be invalid in a variety of ways. your negative testing comment points in that direction. invalid expression formation is caught be the reuse of the expression binding code path, only the 'as' tests should be improved.  testing the alias that follows the 'as' is already covered by the apply code path that is being reused.

---
In reply to: [122828867](https://github.com/OData/odata.net/pull/827#discussion_r122828867) [](ancestors = 122828867)

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ODataUriParserTests.cs [hunk 4]
```diff
@@ -677,7 +677,170 @@ public void ParsePathFunctionWithNullEntitySetPath()
...
+            subPropertyInfo.TargetMultiplicity = EdmMultiplicity.One;
+            EdmProperty subnavigation = EdmNavigationProperty.CreateNavigationProperty(targetType, subPropertyInfo);
+            targetType.AddProperty(subnavigation);
+
+            EdmEntityContainer container = model.AddEntityContainer("Default", "Container");
+            container.AddEntitySet("Entities", elementType);
+
+            // Define queries and new up parser.
+            Uri root = new Uri("http://host");
+            Uri url = new Uri("http://host/Entities?$compute=cast(Prop1, 'Edm.String') as Property1AsString, tolower(Prop1) as Property1Lower&$expand=Nav1($compute=cast(Prop1, 'Edm.String') as NavProperty1AsString;$expand=SubNav1($compute=cast(Prop1, 'Edm.String') as SubNavProperty1AsString))");
```
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ODataUriParserTests.cs]: Can we break this line into multiple lines?

File: src/Microsoft.OData.Core/UriParser/SemanticAst/ExpandedReferenceSelectItem.cs [hunk 5]
```diff
@@ -45,6 +45,33 @@ public ExpandedReferenceSelectItem(ODataExpandPath pathToNavigationProperty, IEd
...
+        /// Create an expand item, using a navigationProperty, its entity set, and any expand options.
+        /// </summary>
+        /// <param name="pathToNavigationProperty">the path to the navigation property for this expand item, including any type segments</param>
+        /// <param name="navigationSource">the navigation source for this expand level.</param>
+        /// <param name="filterOption">A filter clause for this expand (can be null)</param>
+        /// <param name="orderByOption">An Orderby clause for this expand (can be null)</param>
+        /// <param name="topOption">A top clause for this expand (can be null)</param>
+        /// <param name="skipOption">A skip clause for this expand (can be null)</param>
+        /// <param name="countOption">An query count clause for this expand (can be null)</param>
+        /// <param name="searchOption">An levels clause for this expand (can be null)</param>
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/UriParser/SemanticAst/ExpandedReferenceSelectItem.cs]: An levels -> A search

File: src/Microsoft.OData.Core/UriParser/SyntacticAst/ComputeToken.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+        /// <summary>
+        /// Accept a <see cref="ISyntacticTreeVisitor{T}"/> to walk a tree of <see cref="QueryToken"/>s.
+        /// </summary>
+        /// <typeparam name="T">Type that the visitor will return after visiting this token.</typeparam>
+        /// <param name="visitor">An implementation of the visitor interface.</param>
+        /// <returns>An object whose type is determined by the type parameter of the visitor.</returns>
+        public override T Accept<T>(ISyntacticTreeVisitor<T> visitor)
+        {
+            return default(T);
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/UriParser/SyntacticAst/ComputeToken.cs]: Why is this just returning default(T) and not visitor.Visit(this)?

File: src/Microsoft.OData.Core/UriParser/SyntacticAst/ExpandTermToken.cs [hunk 7]
```diff
@@ -104,6 +109,36 @@ public ExpandTermToken(PathSegmentToken pathToNavigationProp, SelectToken select
...
+            PathSegmentToken pathToNavigationProp,
+            QueryToken filterOption,
+            IEnumerable<OrderByToken> orderByOptions,
+            long? topOption,
+            long? skipOption,
+            bool? countQueryOption,
+            long? levelsOption,
+            QueryToken searchOption,
+            SelectToken selectOption,
+            IEnumerable<ComputeToken> computeOptions,
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/UriParser/SyntacticAst/ExpandTermToken.cs]: Why did you add computeOptions *before* expandOption? Generally better for overloads to build on previous parameter ordering.

**Comment by mikepizzo** [comment]
:clock1:

File: src/Microsoft.OData.Core/UriParser/ODataQueryOptionParser.cs [hunk 8]
```diff
@@ -326,6 +331,31 @@ public string ParseDeltaToken()
...
+            if (this.computeClause != null)
+            {
+                return this.computeClause;
+            }
+
+            string computeQuery;
+
+            if (!this.TryGetQueryOption(UriQueryConstants.ComputeQueryOption, out computeQuery)
+                || string.IsNullOrEmpty(computeQuery)
+                || this.targetEdmType == null)
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/UriParser/ODataQueryOptionParser.cs]: >|| this.targetEdmType == null [](start = 16, length = 29)

Under what situations will targetEdmType be null (a dynamic property?) and what are the ramifications of returning a null in this case? Is the compute clause silently ignored? Do we have a test case for this case?

File: src/Microsoft.OData.Core/UriParser/SyntacticAst/ComputeToken.cs (hunk 6 above)
Comment (mtroth-microsoft) [inline src/Microsoft.OData.Core/UriParser/SyntacticAst/ComputeToken.cs]: ISyntacticTreeVisitor<T> is a public interface. It does not have a Visit override that specifically handles ComputeTokens. I opted not to introduce a breaking change since visitation doesn't appear to be required to address the scenarios specified in this issue. I also opted not to introduce a larger non-breaking change for the same reason.

---
In reply to: [123611743](https://github.com/OData/odata.net/pull/827#discussion_r123611743) [](ancestors = 123611743)

File: src/Microsoft.OData.Core/UriParser/ODataQueryOptionParser.cs (hunk 8 above)
Comment (mtroth-microsoft) [inline src/Microsoft.OData.Core/UriParser/ODataQueryOptionParser.cs]: there are two test cases that address returning null, i believe. both seem to be based on the assertion that all query options must be able to handle null scenarios.

---
In reply to: [123615935](https://github.com/OData/odata.net/pull/827#discussion_r123615935) [](ancestors = 123615935)
